### PR TITLE
Using the NLOPT wrapper in the optimization framework (1)

### PR DESCRIPTION
**Pull request by yorkerlin**
@karlnapf

### CODE REVIEW

File: src/shogun/machine/gp/SingleSparseInferenceBase.cpp [hunk 1]
```diff
@@ -270,64 +328,22 @@ void CSingleSparseInferenceBase::optimize_inducing_features()
...
-	// set absolute argument tolearance
-	nlopt_set_xtol_abs1(opt, m_ind_tolerance);
-	nlopt_set_ftol_abs(opt, m_ind_tolerance);
-
-	nlopt_set_min_objective(opt, CSingleSparseInferenceBase::nlopt_function, this);
+	CSingleSparseInferenceCostFunction *cost_fun=new CSingleSparseInferenceCostFunction();
+	SG_REF(this);
+	cost_fun->set_target(this);
+    
+	NLOPTMinimizer* opt=new NLOPTMinimizer(cost_fun);
```
Comment (yorkerlin) [inline src/shogun/machine/gp/SingleSparseInferenceBase.cpp]: @karlnapf 
we can use another optimizer which supports box constraints.

**Comment by yorkerlin** [comment]
merging

File: src/shogun/machine/gp/SingleSparseInferenceBase.cpp [hunk 2]
```diff
@@ -29,23 +29,81 @@
...
+        void set_target(CSingleSparseInferenceBase *obj)
+	{
+		if(obj!=m_obj)
+		{
+			SG_REF(obj);
+			SG_UNREF(m_obj);
+			m_obj=obj;
+			REQUIRE(m_obj,"Object not set\n");
+			m_obj->check_fully_sparse();
+			REQUIRE(m_obj->m_fully_sparse,"Can not compute gradient\n");
```
Comment (karlnapf) [inline src/shogun/machine/gp/SingleSparseInferenceBase.cpp]: Proper error message. State why it cannot do that, what is given and what is expected

Comment (yorkerlin) [inline src/shogun/machine/gp/SingleSparseInferenceBase.cpp]: @karlnapf 
```
void CSingleSparseInferenceBase::check_fully_sparse()
{
        REQUIRE(m_kernel, "Kernel must be set first\n")
        if (strstr(m_kernel->get_name(), "SparseKernel")!=NULL)
                m_fully_sparse=true;
        else
        {
                SG_WARNING( "The provided kernel does not support to optimize inducing features\n");
                m_fully_sparse=false;
        }
}
```

Comment (karlnapf) [inline src/shogun/machine/gp/SingleSparseInferenceBase.cpp]: Why put the warning in the method implicitly?
I would rather `REQUIRE(m_obj->m_fully_sparse,"Can not compute gradient, the provided kernel (%s) does not support to optimize inducing features\n", kernel->get_name());` to be fully explicit

**Pull request merged**